File: templates/minimal-streaming.ts
/**
 * Minimal Streaming Response — Hera Architecture Reference (TypeScript)
 *
 * Pattern: AsyncIterable-based streaming for LLM responses. Enables progressive
 * UI updates, early cancellation, and lower time-to-first-token.
 *
 * Based on Pi Agent's streamAssistantResponse() and Hermes's agent_loop streamFn.
 */

// ============================================================================
// Inline types (self-contained)
// ============================================================================

interface AgentContext {
  systemPrompt: string;
  messages: Array<{ role: string; content: unknown }>;
}

interface ToolCall {
  id: string;
  name: string;
  arguments: unknown;
}

export interface StreamChunk {
  type: "text" | "toolCall" | "usage" | "done";
  text?: string;
  toolCall?: Partial<ToolCall>;
  usage?: { inputTokens: number; outputTokens: number };
}

export interface StreamFn {
  (ctx: AgentContext, signal?: AbortSignal): AsyncIterable<StreamChunk>;
}

// ============================================================================
// Mock streaming provider (replace with real API call)
// ============================================================================

export async function* streamFromProvider(
  ctx: AgentContext,
  signal?: AbortSignal
): AsyncIterable<StreamChunk> {
  // Real implementation: fetch SSE from OpenAI/Anthropic/etc.
  // This is a mock that yields 3 text chunks, 1 tool call, then done.

  const text = "Let me read that file for you.";
  for (const word of text.split(" ")) {
    if (signal?.aborted) return;
    yield { type: "text", text: word + " " };
    await new Promise((r) => setTimeout(r, 50));
  }

  yield {
    type: "toolCall",
    toolCall: {
      id: "call_1",
      name: "read_file",
      arguments: { path: "/repo/src/index.ts" },
    },
  };

  yield { type: "usage", usage: { inputTokens: 100, outputTokens: 12 } };
  yield { type: "done" };
}

// ============================================================================
// Streaming consumer
// ============================================================================

export async function consumeStream(
  stream: AsyncIterable<StreamChunk>,
  onText: (text: string) => void,
  onToolCall: (tc: Partial<ToolCall>) => void,
  signal?: AbortSignal
): Promise<{ toolCalls: Partial<ToolCall>[]; usage?: { inputTokens: number; outputTokens: number } }> {
  const toolCalls: Partial<ToolCall>[] = [];
  let usage: { inputTokens: number; outputTokens: number } | undefined;

  for await (const chunk of stream) {
    if (signal?.aborted) break;

    switch (chunk.type) {
      case "text":
        if (chunk.text) onText(chunk.text);
        break;
      case "toolCall":
        if (chunk.toolCall) {
          toolCalls.push(chunk.toolCall);
          onToolCall(chunk.toolCall);
        }
        break;
      case "usage":
        usage = chunk.usage;
        break;
      case "done":
        return { toolCalls, usage };
    }
  }

  return { toolCalls, usage };
}

File: templates/python/minimal_streaming.py
"""
Minimal Streaming Response — Hera Architecture Reference (Python)

Async generator-based streaming for LLM responses.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import AsyncIterator


# ============================================================================
# Types
# ============================================================================

@dataclass
class StreamChunk:
    type: str  # "text" | "toolCall" | "usage" | "done"
    text: str | None = None
    tool_call: dict | None = None
    usage: dict | None = None


# ============================================================================
# Mock streaming provider
# ============================================================================

async def stream_from_provider(ctx: dict, signal: asyncio.Event | None = None) -> AsyncIterator[StreamChunk]:
    """Mock provider that yields text chunks, a tool call, then done."""
    text = "Let me read that file for you."
    for word in text.split(" "):
        if signal and signal.is_set():
            return
        yield StreamChunk(type="text", text=word + " ")
        await asyncio.sleep(0.05)

    yield StreamChunk(
        type="toolCall",
        tool_call={
            "id": "call_1",
            "name": "read_file",
            "arguments": {"path": "/repo/src/index.ts"},
        },
    )
    yield StreamChunk(type="usage", usage={"input_tokens": 100, "output_tokens": 12})
    yield StreamChunk(type="done")


# ============================================================================
# Consumer
# ============================================================================

async def consume_stream(
    stream: AsyncIterator[StreamChunk],
    on_text,
    on_tool_call,
    signal: asyncio.Event | None = None,
) -> dict:
    """Consume a stream, calling callbacks for each chunk."""
    tool_calls: list[dict] = []
    usage: dict | None = None

    async for chunk in stream:
        if signal and signal.is_set():
            break

        if chunk.type == "text" and chunk.text:
            on_text(chunk.text)
        elif chunk.type == "toolCall" and chunk.tool_call:
            tool_calls.append(chunk.tool_call)
            on_tool_call(chunk.tool_call)
        elif chunk.type == "usage":
            usage = chunk.usage
        elif chunk.type == "done":
            return {"tool_calls": tool_calls, "usage": usage}

    return {"tool_calls": tool_calls, "usage": usage}
